src/components/graph/ExerciseGraphLayout.ts

```diff
@@ -0,0 +1,337 @@
+        if (depth > maxDepths[index]) {
+          maxDepths[index] = depth
+        }
+
+        const children = this.adjacency[index]
+        children.forEach((child) => traverse(child, depth + 1))
+      }
+      traverse(source)
+    }
+    return maxDepths
+  }
+
+  /**
+   * order the exercises by depth then index into layers
+   */
+  private constructExerciseLayers(
+    exerciseDepths: Array<number>,
+    maxDepth: number
+  ): Array<Array<ExerciseGraphNode>> {
+    const exerciseLayers = new Array(maxDepth + 1)
+
+    exerciseDepths.forEach((nodeDepth, nodeIndex) => {
+      exerciseLayers[nodeDepth] = exerciseLayers[nodeDepth] ?? []
+      exerciseLayers[nodeDepth].push(this.graph.lookupByIndex.get(nodeIndex))
+    })
+
+    return exerciseLayers
+  }
+
+  /**
+   * order the concepts without exercises by the depth of their dependent exercise
+   */
+  private constructConceptLayers(
+    exerciseLayers: Array<Array<ExerciseGraphNode>>
+  ): Array<Array<string>> {
+    // create map for layer position for a missing concept in case the concept is listed
+    // by multiple exercises
+    const conceptsVisited = new Map<string, boolean>()
+
+    // compute missing concept layers
+    const conceptLayers = new Array<Array<string>>()
+    exerciseLayers.forEach((layer, i) => {
+      conceptLayers.push([])
+      layer.forEach((node) => {
+        const missingPrereqs = this.graph.lookupMissingConceptsForExercise.get(
+          node.slug
+        )
+        if (!missingPrereqs) return
+
+        missingPrereqs.forEach((prereq) => {
+          const visited = conceptsVisited.get(prereq)
+          if (visited) return
+          conceptLayers[i].push(prereq)
+          conceptsVisited.set(prereq, true)
+        })
+      })
+    })
+
+    return conceptLayers
+  }
+
+  private computePositions(): void {
+    /**
+     * Compute amount to offset each exercise node from the left for each row
+     * Also, if the rows have the same number of exercises, slightly offset each one
+     * so that the dependency lines aren't completely straight/overlapping
+     */
+    const xAdjust = -40
+    const exerciseXOffset = new Array(this.exerciseLayerCount)
+    const exerciseXAdjust = new Array(this.exerciseLayerCount).fill(0)
+    this.exerciseLayers.forEach((layer, i) => {
+      exerciseXOffset[i] = Math.floor(
+        (this.width - this.padding * 2) / (layer.length + 1)
+      )
+
+      const prevLayer = this.exerciseLayers[i - 1]
+      if (prevLayer && prevLayer.length === layer.length) {
+        exerciseXAdjust[i] = exerciseXAdjust[i - 1] + 5
+      }
+    })
+
+    /**
+     * Compute amount to offset each node from the left
+     */
+    const conceptXAdjust = -25
+    const conceptXOffset = new Array(this.conceptLayers.length)
+    this.conceptLayers.forEach((layer, i) => {
+      if (layer.length === 0) return
+
+      conceptXOffset[i] = Math.floor(
+        (this.width - this.padding * 2) / (layer.length + 1)
+      )
+    })
```

Comment:
```suggestion
    const conceptXOffset = new Array(this.conceptLayers.length)
    this.conceptLayers.forEach((layer, i) => {
      if (layer.length === 0) {
        return
      }

      conceptXOffset[i] = Math.floor(
        (this.width - this.padding * 2) / (layer.length + 1)
      )
    })
```

But I think you actually wanted:

```suggestion
    const conceptXOffset = this.conceptLayers.map((layer, i) =>
      return layer.length === 0
        ? 0
        : Math.floor((this.width - this.padding * 2) / (layer.length + 1))
    )
```